File: Problems/0023_Merge_k_Sorted_Lists/Project_Python3/Merge_k_Sorted_Lists.py
import os
import sys
import time

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def mergeKLists(self, lists: 'List[ListNode]') -> 'ListNode':
        ans = []
        for ln in lists:
            while ln:
                ans.append(ln.val)
                ln = ln.next
        ans.sort()
        result = set_nodes(ans, 0)

        return result

    def mergeKLists_work(self, lists: 'List[ListNode]') -> 'ListNode':
        if len(lists) < 1:
            return None
        if lists == None:
            return None

        none_count = 0
        for temp_list in lists:
            if temp_list == None:
                none_count += 1
        if none_count == len(lists):
            return None

        current_lists = lists
        work_list = None
        while True:
            current_min = sys.maxsize
            end_count = 0
            for i in range(len(current_lists)):
                if current_lists[i] == None:
                    end_count += 1
                    continue
                if current_lists[i].val < current_min:
                    current_index = i
                    current_min = current_lists[i].val
            if end_count == len(current_lists):
                break

            if work_list == None:
                root = ListNode(current_min)
                work_list = root
            else:
                work_list.next = ListNode(current_min)
                work_list = work_list.next

            current_lists[current_index] = current_lists[current_index].next

        return root

def str_to_int_array(flds):
    if len(flds) <= 0:
        return None
    temp = flds.split(",")
    nums = [0]*len(temp)
    for i in range(len(temp)):
        nums[i] = int(temp[i])
    return nums

def set_nodes(nums, index):
    if nums == None:
        return None
    if index >= len(nums):
        return None
    
    node = ListNode(nums[index])
    node.next = set_nodes(nums, index + 1)

    return node

def output_nodes(ll):
    if ll == None:
        return ""

    retStr = str(ll.val)
    if ll.next != None:
        retStr += " -> " + output_nodes(ll.next)
    return retStr

def main():
    argv = sys.argv
    argc = len(argv)

    if (argc < 2):
        print("Usage: python %s <testdata.txt>" %(argv[0]))
        exit(0)

    if not os.path.exists(argv[1]):
        print("%s not found..." %argv[1])
        exit(0)

    testDataFile = open(argv[1], "r")
    lines = testDataFile.readlines()

    for temp in lines:
        temp = temp.strip()
        if temp == "":
            continue
        print("args = %s" %temp)
        loop_main(temp)
    #    print("Hit Return to continue...")
    #    input()

def loop_main(temp):
    str_args = temp.replace("\"","").replace("[[","").replace("]]","").rstrip()
    flds = str_args.split("],[")

    nums = [0]*len(flds)
    for i in range(len(flds)):
        nums[i] = str_to_int_array(flds[i])
        print("nums[%d] = %s " %(i, nums[i]))

    lists = []
    for i in range(len(nums)):
        lists.append(set_nodes(nums[i], 0))
        print("lists[%d] = %s " %(i, output_nodes(lists[i])))

    time0 = time.time()

    sl = Solution()
    result = sl.mergeKLists(lists)

    time1 = time.time()
    print("result = %s" %output_nodes(result))
    print("Execute time ... : %f[s]\n" %(time1 - time0))

if __name__ == "__main__":
    main()

File: Problems/0001_0099/0023_Merge_k_Sorted_Lists/Project_CS/Merge_k_Sorted_Lists.cs
using System;

//Definition for singly-linked list.
public class ListNode {
    public int val;
    public ListNode next;
    public ListNode(int x) { val = x; }
}

public class Solution
{
    public ListNode MergeKLists(ListNode[] lists)
    {
        if (lists.Length < 1)
            return null;
        if (lists == null)
            return null;

        int none_count = 0;
        foreach (ListNode temp_list in lists)
        {
            if (temp_list == null)
                none_count++;
        }
        if (none_count == lists.Length)
            return null;

        ListNode root = null;
        ListNode node = null;

        foreach (ListNode target_list in lists)
        {
            ListNode work_list = target_list;

            while (work_list != null)
            {
                if (node == null)
                {
                    root = new ListNode(work_list.val);
                    node = root;
                }
                else
                {
                    node.next = new ListNode(work_list.val);
                    node = node.next;
                }

                work_list = work_list.next;
            }
        }

        return Sort_ListNode(root);
    }

    public ListNode Sort_ListNode(ListNode list)
    {
        ListNode root = list;
        ListNode target = null;
        int temp;

        ListNode work = root;
        while (work != null)
        {
            target = work.next;
            while (target != null)
            {
                if (work.val > target.val)
                {
                    temp = work.val;
                    work.val = target.val;
                    target.val = temp;
                }
                target = target.next;
            }
            work = work.next;
        }

        return root;
    }

    public int[] str_to_int_array(string s)
    {
        if (s.Length <= 0)
            return null;

        string[] flds = s.Split(',');
        int[] nums = new int[flds.Length];

        if (flds.Length <= 0)
            return nums;

        for (int i = 0; i < nums.Length; ++i)
        {
            nums[i] = int.Parse(flds[i]);
        }

        return nums;
    }

    public string output_int_array(int[] nums)
    {
        if (nums == null)
            return "";

        if (nums.Length <= 0)
            return "";

        string resultStr = "[" +  nums[0].ToString();
 
        for (int i = 1; i < nums.Length; ++i)
        {
            resultStr += "," + nums[i].ToString();
        }

        return resultStr + "]";
    }

    public void Main(string args)
    {
        string arg_str = args.Replace("[[","").Replace("]]","").Trim();
        string[] flds = arg_str.Split(new string[] {"],["}, StringSplitOptions.None);

        int[][] nums = new int[flds.Length][];
        for (int i = 0; i < nums.Length; ++i)
        {
            nums[i] = str_to_int_array(flds[i]);
            Console.WriteLine("nums[" + i.ToString() + "] = " + output_int_array(nums[i]));
        }

        Operate_ListNode ope = new Operate_ListNode();
        ListNode[] lists = new ListNode[nums.Length];

        for (int i = 0; i < nums.Length; ++i)
        {
            lists[i] = ope.set_ListNode(nums[i]);
            Console.WriteLine("lists[" + i.ToString() + "] = " + ope.output_ListNode(lists[i]));
        }

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        ListNode result = MergeKLists(lists);
        sw.Stop();

        Console.WriteLine("result = " + ope.output_ListNode(result));
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms\n");
    }
}
